Improve Rviz plugin usability and performance

victorreijgwart opened the pull request:
This PR proposes a full rewrite of the Rviz plugin, to improve performance and ease of use.

New features:
* Communication
  * Incremental transmission of the map (avoids exceeding ROS message size limits for very large maps)
  * Rewritten map<->ROS msg conversions, matching the new map<->byte stream implementation
* Rviz plugin UI
  * Support different render modes (slice; grid) in a single WavemapDisplay instance
    * Each render mode can be configured through its own foldout menu
    * The map is only stored once per plugin instance and shared among the render mode handlers
  * Grid render mode
    * Expose grid color options in the UI
    * Add option to set maximum grid drawing resolution
      * E.g. to improve frame rate when displaying large maps on computers with no GPU
  * Improve default settings s.t. it can be used with minimal tuning in most scenarios
* Performance improvements
  * Only redraw map blocks that changed
  * Render grid blocks with Level of Detail based on their distance to the camera
  * Use a work queue and limit the update time per frame, to avoid stalling Rviz when large map changes occur
  * Interface directly with Ogre, instead of using rviz::Pointcloud as an intermediary for rendering

Removed features:
* Remove the option to render map meshes
  * This render mode does not yet produce good iso-surfaces and is currently very slow. It will be reintroduced once its more mature.

LionelOtt commented on libraries/wavemap/include/wavemap/utils/time.h:
```diff
@@ -0,0 +1,17 @@
+#include <chrono>
+
+namespace wavemap {
+using Time = std::chrono::steady_clock;
```
I assume steady over high_resolution so you're certain time doesn't run backwards?

LionelOtt commented on ros/wavemap_ros/launch/online/pico_monstar.launch:
```diff
@@ -83,6 +83,6 @@
-<!--  <node pkg="rosbag" type="record" name="rosbag_record" output="screen"-->
-<!--        args="/wavemap/depth_cam_input /laserMapping/imu_forward_prop -O /home/victor/catkin_ws/lidar_align_pico_monstar"/>-->
+  <node if="false" pkg="rosbag" type="record" name="rosbag_record" output="screen"
+        args="/livox/lidar /livox/imu /pico_monstar/image_rect -o /home/victor/catkin_ws/livox_monstar"/>
```
Could this absolute user-specific path be changed to something more general based on the current WS or similar?

victorreijgwart commented:
Testing with GCC's Thread Sanitizer is currently failing in GitHub actions, as the linker fails to find `libtsan_preinit.o`. The file seems to have disappeared in a recent update. The issue is reproducible on local machines running 20.04 with up-to-date apt packages.

The issue matches this Ubuntu bug report:
https://bugs.launchpad.net/ubuntu/+source/gcc-10/+bug/2029910

Hopefully, it will be fixed soon. Otherwise the best option is probably to ignore TSAN and move ahead with the merge once I addressed your comments.

victorreijgwart commented on libraries/wavemap/include/wavemap/data_structure/volumetric/hashed_chunked_wavelet_octree_block.h:
```diff
@@ -57,10 +54,10 @@ class HashedChunkedWaveletOctreeBlock {
+  void setLastUpdatedStamp(Timestamp stamp = Time::now()) {
     last_updated_stamp_ = stamp;
   }
-  Time getLastUpdatedStamp() const { return last_updated_stamp_; }
```
Short answer: yes :)
Long answer: This way of handling time internally in the framework works well with ROS time (incl. bag playback) in the sense that the two time sources are never mixed, converted, or compared.
* ROS time is only used in the ROS interface (wavemap_server) to convert pointclouds+timestamps+odometry --> pointcloud+pose.
* The library internally uses temporal information only when checking if the map changed since some event occurred (e.g. since it was last published) and always does this using the same clock, accessed through the `Time = std::chrono::steady_clock;` alias for consistency.

victorreijgwart commented:
/prepare-release minor

victorreijgwart commented:
> Have a quick look over the comments I've added but otherwise looks good to me.

Thanks for reviewing!
